ccmlib/common.py

```diff
@@ -943,20 +943,25 @@ def assert_jdk_valid_for_cassandra_version(cassandra_version):
         exit(1)
 
 
-def aws_bucket_ls(s3_url):
+def aws_bucket_ls(s3_url, public=True):
     bucket_object = s3_url.replace('https://s3.amazonaws.com/', '').split('/')
     prefix = '/'.join(bucket_object[1:])
-    s3_conn = Session().client(service_name='s3', config=Config(signature_version=UNSIGNED))
+    config = Config(signature_version=UNSIGNED) if public else None
+    s3_conn = Session().client(service_name='s3', config=config)
     paginator = s3_conn.get_paginator('list_objects_v2')
     pages = paginator.paginate(Bucket=bucket_object[0], Prefix=prefix)
 
+    def get_last_modified(obj):
+        return obj['LastModified']
+
     files_in_bucket = []
     for page in pages:
         if 'Contents' not in page:
             break
 
-        for obj in page['Contents']:
-            files_in_bucket.append(obj['Key'].replace(prefix + "/", ''))
+        files_in_bucket.extend(page['Contents'])
+
+    files_in_bucket = [f['Key'].replace(prefix + "/", '') for f in sorted(files_in_bucket, key=get_last_modified)]
```

Comment:
i'd recommend to avoid assigning different type of data to the same variable. after this change, `files_in_bucket` is used in two places:

1. to hold the objects listed with given prefix: it is a list of dict
2. to hold their keys after being normalized: it is a list of str

this would add the cognitive load of human readers. i'd suggest use something like
```py
for page in pages:
    for obj in page.get('Contents', []):
        files_in_bucket.append(obj['Key'].replace(prefix + "/", ''))
return sorted(files_in_bucket, key=get_last_modified)
```

[offtopic], i am wondering if we can restructure the code using Collections. like
```py
bucket, prefix = s3_url.replace('https://s3.amazonaws.com/', '').split('/', 1)
s3 = boto3.resource('s3')
objects_in_bucket = (obj.key.replace(prefix + "/", '') for s3.Bucket(bucket).objects.all())
return sorted(objects_in_bucket, key=lambda obj: obj.last_modified)
```
where we can have some improvements:

1. construct the prefix when parsing the `s3_url`
1. reuse the builtin support of resource management
1. do not paginate the list_object_v2 manually
1. use iterator instead of materializing the list when possible
1. access the attribute of enumerated object instead of accessing it like a dict

see https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#list-objects-in-an-amazon-s3-bucket and https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/object/last_modified.html#last-modified